Migrate to GHA

claire-labry opened the pull request:
This PR officially migrates `go-getter` to GHA from CircleCI. It's intention is to work the same as the CircleCI tests but in GHA format. 

Note: After working on updating the AWS/GCP tests, it was concluded that the older go versions (1.14.15 and 1.15.3) were blocking the GCP tests from passing. The issue was that go1.17 is required for the dependent GCS libraries and we concluded that the removal of the older go versions was the logical path to move forward.

picatz commented:
Just to be clear, we will disable the CircleCI pipeline _after_ this PR is merged? 🤔

claire-labry commented:
Yes, that's the intention! @picatz

picatz commented on get_s3_test.go:
```diff
@@ -31,7 +31,7 @@ func TestS3Getter(t *testing.T) {
 
```
We should probably remove the [`init`](https://github.com/hashicorp/go-getter/blob/d305097eb8901ba73041f9b4efcc8eae90f15dcd/get_s3_test.go#L12-L22) from this file, and document the expectation for how these tests should run (only in CI from PRs). This will be useful in the future, where external contributors will not be able to run these tests locally anymore, for example.
https://github.com/hashicorp/go-getter/blob/d305097eb8901ba73041f9b4efcc8eae90f15dcd/get_s3_test.go#L12-L22

picatz commented on get_gcs_test.go:
```diff
@@ -8,35 +8,17 @@ import (
 	"testing"
 )
 
-// initGCPCredentials writes a temporary GCS credentials file if necessary and
```
Similar to my other comment for S3, we should documentation the expectation for where/how these tests should run. It won't be obvious if someone tries to run the tests locally for this module.

picatz commented on .github/workflows/go-getter.yml:
```diff
@@ -0,0 +1,156 @@
+      
+      - name: Install gotestsum
+        run: go install gotest.tools/gotestsum@v1.8.2
+
```
Optional formatting nit:

```suggestion
```

picatz commented on .github/workflows/go-getter.yml:
```diff
@@ -0,0 +1,156 @@
+        shell: bash
+        run: go install gotest.tools/gotestsum@v1.8.2
+        
+
```
Optional formatting nit:

```suggestion
```

picatz commented on .github/workflows/go-getter.yml:
```diff
@@ -0,0 +1,156 @@
+      - name: Download go modules
+        run: go mod download
+      
+      - name: Cache / restore go modules
```
Optional consistency nit, different name from the other step name for Windows tests:

```suggestion
      - name: Setup cache for go modules
```

picatz commented on .github/workflows/go-getter.yml:
```diff
@@ -0,0 +1,156 @@
+    strategy:
+      matrix:
+        go-version:
+          - 1.19
```
I suspect we'd be able to test `1.17`, `1.18`, and `1.19`, what do you think? 🤔 

```suggestion
        go-version:
          - 1.17
          - 1.18
          - 1.19
```

picatz commented on .github/workflows/go-getter.yml:
```diff
@@ -0,0 +1,156 @@
+    runs-on: windows-latest
+    strategy:
+      matrix:
+        go-version:
+          - 1.19
```
Similar to my suggestion in the Linux test steps, I suspect we'd be able to test `1.17`, `1.18`, and `1.19`?

```suggestion
        go-version:
          - 1.17
          - 1.18
          - 1.19
```

claire-labry commented:
I'm going to defer to @dpowley for the first two comments on the s3 tests?

samsalisbury reviewed (commented):
Thanks Claire, the caching seems to be broken atm, I left comments below... Other than that the CI bits all LGTM!

claire-labry commented on .github/workflows/go-getter.yml:
```diff
@@ -0,0 +1,158 @@
+      - name: Download go modules
+        run: go mod download
+      
+      - name: Setup cache for go modules
+        uses: actions/cache@v3
+        with:
+          path: |
+            ~/.cache/go-build
+            ~/go/pkg/mod
+          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-
```
ah yeah that's my bad-- thank you for this tip!

claire-labry commented:
@samsalisbury @picatz -- Can I get a re-review on this PR? The cache tests have been fixed -- @nywilken Would it be ok if we migrated [v2](https://github.com/hashicorp/go-getter/tree/v2) to GHA?

claire-labry commented:
@picatz would you be able to re-review this PR as I've added several things since your last review: [this commit here](https://github.com/hashicorp/go-getter/pull/379/commits/bad826d9c545eb209425c991226c487b0093641d) -- we need to allow the file to be transported globally due to the CVE linked in the [comments](https://github.com/hashicorp/go-getter/pull/379/commits/47ecbb0e0aae24d0304b71bea95a0f89153e0d2f). I also added comments to the GCS and S3 tests for external contributors on how they are able to run these tests. Once I have your final review and approval, I will merge this in.

picatz commented on get_s3_test.go:
```diff
@@ -9,17 +9,8 @@ import (
-	os.Setenv("AWS_ACCESS_KEY", "AKIAITTDR"+"WY2STXOZE2A")
-	os.Setenv("AWS_SECRET_KEY", "oMwSyqdass2kPF"+"/7ORZA9dlb/iegz+89B0Cy01Ea")
-}
+//Note for external contributors: In order to run the GCS test suite, you will only be able to be run
```
Small formatting nit and typo fix.
```suggestion
// Note for external contributors: In order to run the S3 test suite, you will only be able to be run
```

claire-labry replied:
🤦🏻‍♀️ this is why i need more coffee! thank you

picatz replied:
😄 ☕

picatz commented on get_gcs_test.go:
```diff
@@ -8,35 +8,20 @@ import (
-	}
-	return func() {}
-}
+//Note for external contributors: In order to run the GCS test suite, you will only be able to be run
```
Small formatting nit.
```suggestion
// Note for external contributors: In order to run the GCS test suite, you will only be able to be run
```
